app/controllers/api/storage.php

```diff
@@ -306,97 +307,200 @@
-            'name' => $file['name'],
-            'path' => $path,
-            'signature' => $device->getFileHash($path),
-            'mimeType' => $mimeType,
-            'sizeOriginal' => $size,
-            'sizeActual' => $sizeActual,
-            'algorithm' => empty($compressor) ? '' : $compressor->getName(),
-            'comment' => '',
-        ];
-
-        if($bucket->getAttribute('encryption', true) && $size <= APP_LIMIT_ENCRYPTION) {
-            $data['openSSLVersion'] = '1';
-            $data['openSSLCipher'] = OpenSSL::CIPHER_AES_128_GCM;
-            $data['openSSLTag'] = \bin2hex($tag);
-            $data['openSSLIV'] = \bin2hex($iv);
-        }
-
-        $file = $dbForInternal->createDocument('files', new Document($data));
+            if (App::getEnv('_APP_STORAGE_ANTIVIRUS') === 'enabled' && $bucket->getAttribute('antiVirus', true) && $size <= APP_LIMIT_ANTIVIRUS) {
+                $antiVirus = new Network(App::getEnv('_APP_STORAGE_ANTIVIRUS_HOST', 'clamav'),
+                    (int) App::getEnv('_APP_STORAGE_ANTIVIRUS_PORT', 3310));
+
+                if (!$antiVirus->fileScan($path)) {
+                    $device->delete($path);
+                    throw new Exception('Invalid file', 403);
+                }
+            }
+
+            // Compression
+            if($size <= APP_LIMIT_COMPRESSION || $size <= APP_LIMIT_ENCRYPTION) {
+                $data = $device->read($path);
+                if ($size <= APP_LIMIT_COMPRESSION) {
+                    $compressor = new GZIP();
+                    $data = $compressor->compress($data);
+                }
+    
+                if ($bucket->getAttribute('encryption', true) && $size <= APP_LIMIT_ENCRYPTION) {
+                    $key = App::getEnv('_APP_OPENSSL_KEY_V1');
+                    $iv = OpenSSL::randomPseudoBytes(OpenSSL::cipherIVLength(OpenSSL::CIPHER_AES_128_GCM));
+                    $data = OpenSSL::encrypt($data, OpenSSL::CIPHER_AES_128_GCM, $key, 0, $iv, $tag);
+                }
+                if (!$device->write($path, $data, $mimeType)) {
+                    throw new Exception('Failed to save file', 500);
+                }
+            }
+
+
+            $sizeActual = $device->getFileSize($path);
+
+            $read = (is_null($read) && !$user->isEmpty()) ? ['user:' . $user->getId()] : $read ?? [];
+            $write = (is_null($write) && !$user->isEmpty()) ? ['user:' . $user->getId()] : $write ?? [];
+            $algorithm = empty($compressor) ? '' : $compressor->getName();
+            $fileHash = $device->getFileHash($path);
+
+            if ($bucket->getAttribute('encryption', true) && $size <= APP_LIMIT_ENCRYPTION) {
+                $openSSLVersion = '1';
+                $openSSLCipher = OpenSSL::CIPHER_AES_128_GCM;
+                $openSSLTag = \bin2hex($tag);
+                $openSSLIV = \bin2hex($iv);
+            }
+
+            if ($file->isEmpty()) {
+                $data = [
```

Comment:
You don't really need this var. You can just create the document instance right here.